File: docs/snippets/all/views/timeseries.py
"""Use a blueprint to customize a TimeSeriesView."""

import math

import rerun as rr
import rerun.blueprint as rrb

rr.init("rerun_example_timeseries", spawn=True)

# Log some trigonometric functions
rr.log("trig/sin", rr.SeriesLines(colors=[255, 0, 0], names="sin(0.01t)"), static=True)
rr.log("trig/cos", rr.SeriesLines(colors=[0, 255, 0], names="cos(0.01t)"), static=True)
rr.log("trig/cos_scaled", rr.SeriesLines(colors=[0, 0, 255], names="cos(0.01t) scaled"), static=True)
for t in range(int(math.pi * 4 * 100.0)):
    rr.set_time("timeline0", sequence=t)
    rr.set_time("timeline1", duration=t)
    rr.log("trig/sin", rr.Scalars(math.sin(float(t) / 100.0)))
    rr.log("trig/cos", rr.Scalars(math.cos(float(t) / 100.0)))
    rr.log("trig/cos_scaled", rr.Scalars(math.cos(float(t) / 100.0) * 2.0))

# Create a TimeSeries View
blueprint = rrb.Blueprint(
    rrb.Vertical(
        contents=[
            rrb.TimeSeriesView(
                origin="/trig",
                # Set a custom Y axis.
                axis_y=rrb.ScalarAxis(range=(-1.0, 1.0), zoom_lock=True),
                # Configure the legend.
                plot_legend=rrb.PlotLegend(visible=False),
                # Set time different time ranges for different timelines.
                time_ranges=[
                    # Sliding window depending on the time cursor for the first timeline.
                    rrb.VisibleTimeRange(
                        "timeline0",
                        start=rrb.TimeRangeBoundary.cursor_relative(seq=-100),
                        end=rrb.TimeRangeBoundary.cursor_relative(),
                    ),
                    # Time range from some point to the end of the timeline for the second timeline.
                    rrb.VisibleTimeRange(
                        "timeline1",
                        start=rrb.TimeRangeBoundary.absolute(seconds=300.0),
                        end=rrb.TimeRangeBoundary.infinite(),
                    ),
                ],
            ),
            rrb.TimeSeriesView(
                origin="/trig",
                axis_x=rrb.TimeAxis(
                    view_range=rr.TimeRange(
                        start=rrb.TimeRangeBoundary.cursor_relative(seconds=-100),
                        end=rrb.TimeRangeBoundary.cursor_relative(seconds=100),
                    ),
                    zoom_lock=True,
                ),
                # Configure the legend.
                plot_legend=rrb.PlotLegend(visible=True),
                background=rrb.archetypes.PlotBackground(color=[128, 128, 128], show_grid=False),
            ),
        ]
    ),
    collapse_panels=True,
)

rr.send_blueprint(blueprint)

File: docs/snippets/all/views/timeseries.rs
//! Use a blueprint to customize a TimeSeriesView.

use rerun::blueprint::{
    Blueprint, TimeSeriesView, Vertical, archetypes as blueprint_archetypes,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let blueprint = Blueprint::new(Vertical::new([
        TimeSeriesView::new("Y axis and time ranges")
            .with_origin("/trig")
            .with_axis_y(
                blueprint_archetypes::ScalarAxis::new()
                    .with_range([-1.0, 1.0])
                    .with_zoom_lock(true),
            )
            .with_plot_legend(
                blueprint_archetypes::PlotLegend::new().with_visible(false),
            )
            .with_time_ranges(blueprint_archetypes::VisibleTimeRanges::new([
                rerun::datatypes::VisibleTimeRange {
                    timeline: "timeline0".into(),
                    range: rerun::datatypes::TimeRange {
                        start:
                            rerun::datatypes::TimeRangeBoundary::CursorRelative(
                                (-100).into(),
                            ),
                        end: rerun::datatypes::TimeRangeBoundary::AT_CURSOR,
                    },
                },
                rerun::datatypes::VisibleTimeRange {
                    timeline: "timeline1".into(),
                    range: rerun::datatypes::TimeRange {
                        start: rerun::datatypes::TimeRangeBoundary::Absolute(
                            300_000_000_000_i64.into(),
                        ),
                        end: rerun::datatypes::TimeRangeBoundary::Infinite,
                    },
                },
            ]))
            .into(),
        TimeSeriesView::new("X axis and background")
            .with_origin("/trig")
            .with_axis_x(
                blueprint_archetypes::TimeAxis::new()
                    .with_view_range(
                        rerun::datatypes::TimeRange::from_cursor_plus_minus(
                            100,
                        ),
                    )
                    .with_zoom_lock(true),
            )
            .with_plot_legend(
                blueprint_archetypes::PlotLegend::new().with_visible(true),
            )
            .with_background(
                blueprint_archetypes::PlotBackground::new()
                    .with_color([128, 128, 128])
                    .with_show_grid(false),
            )
            .into(),
    ]));

    let rec = rerun::RecordingStreamBuilder::new("rerun_example_timeseries")
        .with_blueprint(blueprint)
        .spawn()?;

    rec.log_static(
        "trig/sin",
        &rerun::SeriesLines::new()
            .with_colors([[255, 0, 0]])
            .with_names(["sin(0.01t)"]),
    )?;
    rec.log_static(
        "trig/cos",
        &rerun::SeriesLines::new()
            .with_colors([[0, 255, 0]])
            .with_names(["cos(0.01t)"]),
    )?;
    rec.log_static(
        "trig/cos_scaled",
        &rerun::SeriesLines::new()
            .with_colors([[0, 0, 255]])
            .with_names(["cos(0.01t) scaled"]),
    )?;

    for t in 0..((std::f64::consts::PI * 4.0 * 100.0) as i64) {
        rec.set_time_sequence("timeline0", t);
        rec.set_duration_secs("timeline1", t as f64);
        rec.log(
            "trig/sin",
            &rerun::Scalars::single((t as f64 / 100.0).sin()),
        )?;
        rec.log(
            "trig/cos",
            &rerun::Scalars::single((t as f64 / 100.0).cos()),
        )?;
        rec.log(
            "trig/cos_scaled",
            &rerun::Scalars::single((t as f64 / 100.0).cos() * 2.0),
        )?;
    }

    Ok(())
}
